### DISCUSSION TITLE
Explicit self referenced many-to-many relationship

### DISCUSSION THREAD

**Discussion by TheRealThor** [prisma/prisma#6104] [Q&A] [answered]
Hello guys,

that's an impressive work you guys have done. I am coming from a PHP/Laravel background, where SQL is as well abstracted with the famous[ Eloquent Relationships](https://laravel.com/docs/8.x/eloquent-relationships) of that framework. 
I am thinking of switching to nodejs with  Prisma.  However when I was playing with the Prisma Schema I stumbled on an issue.  Per the documentation implicit relationships are possible with the following notation

```prisma
model User {
  id          Int      @id @default(autoincrement())
  name        String?
  followedBy  User[]   @relation("UserFollows", references: [id])
  following   User[]   @relation("UserFollows", references: [id])
}
```

I din't find a way to change that to an **explicit relationship** where I can add columns/fields to describe that relationship further. I tried to define an extra model to represent a relation table:

```prisma
 model User {
  id          Int      @id @default(autoincrement())
  name        String?
  followedBy  UserOnUser[]   
  following   UserOnUser[]   
}

model UserOnUser {
    child   User @relation(fields: [childId], references: [id])
    childId Int

    parent   User @relation(fields: [parentId], references: [id])
    parentId Int

   following_type  String

   @@id([childId, parentId])

}
```

But this did not work  and I got the following errors: 

```
Error validating model "User": Ambiguous relation detected. The fields `children` and `parents` in model `Person` both refer to `PersonOnPerson`. Please provide different relation names for them by adding `@relation(<name>).

Error validating model "UserOnUser": Ambiguous relation detected. The fields `child` and `parent` in model `PersonOnPerson` both refer to `Person`. Please provide different relation names for them by adding `@relation(<name>).
```

Is it possible to define explicit self referenced many-to-many relationships with Prisma (yet)?

**Comment by ryands17** [comment, marked as answer]
@TheRealThor 👋 

You need to disambiguate relations whenever you have two or more relations pointing to the same model. So your schema will look like:

```prisma
model User {
  id         Int          @id @default(autoincrement())
  name       String?
  followedBy UserOnUser[] @relation("followedBy")
  following  UserOnUser[] @relation("following")
}

model UserOnUser {
  child   User @relation("following", fields: [childId], references: [id])
  childId Int

  parent   User @relation("followedBy", fields: [parentId], references: [id])
  parentId Int

  following_type String


  @@id([childId, parentId])
}
```

You can read more about this [here](https://www.prisma.io/docs/concepts/components/prisma-schema/relations#disambiguating-relations).

**Reply by Dammmien** [reply]
Hi, I'm facing a similar issue and when I'm querying the collection it returns me the `UserOnUser` items and it doesn't map it to the `User` itself. I have something like that:

```ts
const list = await prisma.users.findMany({
  include: {
    following: true
  }
});

console.log(list[0].following[0])
// { childId: 1, parentId: 2 }
```

I have a `UserOnUser` object but I want a `User` one.
It works with nested includes:

```ts
const list = await prisma.users.findMany({
  include: {
    following: {
      include: {
        child: true
      }
    }
  }
});

console.log(list[0].following[0])
// { childId: 1, parentId: 2, child: { .... } }
```

But it's not very convenient, to have `list[0].following[0].child` and I have the feeling it should not work like that. What do I need to change ?

**Reply by Dammmien** [reply]
In fact it seems to be the expected behavior: https://www.prisma.io/docs/support/help-articles/working-with-many-to-many-relations#explicit-relations and we have to manually map the list. Not very convenient but I can understand the need to not map automatically allowing to retrieve information from the relation.

**Reply by faustinozanetto** [reply]
Hey I was having a similar issue and it solved it, thanks. Now I would like to now how can I make a user follow another one using their ids. I tried using prisma.user.update but no luck. Thanks in advance.

**Reply by ryands17** [reply]
@faustinozanetto Could you share what you passed in `prisma.user.update`?

**Reply by faustinozanetto** [reply]
Hi, thanks for replaying.
I ended up using prisma.userOnUser.create after several trials and errors though I don´t know if it is the best way to do it.

```typescript
import { NextApiRequest, NextApiResponse } from 'next';
import prisma from '@lib/prisma/prisma';

export default async (req: NextApiRequest, res: NextApiResponse) => {
  try {
    // Getting current user id and the user to follow id from the req body.
    const { id, targetId } = req.body;
    // Creating an userOnUser with the ids.
    const followed = await prisma.userOnUser.create({
      data: {
        parentId: id,
        childId: targetId,
      },
    });
    // Returning the result
    res.status(200);
    res.json(followed);
  } catch (error) {
    // Error handling.
    console.error(error);
    res.status(500);
    res.json({ error: 'Sorry unable to save sighting to database' });
  }
};
```

**Reply by ryands17** [reply]
This is fine as well.

**Reply by faustinozanetto** [reply]
Thanks! The only thing that I am aware of is that user can see the api and be able to call the api route with users id and be able to follow users.

**Reply by ryands17** [reply]
I didn't quite get your point. Could you explain?

**Reply by faustinozanetto** [reply]
I am currently using just the frontend nextjs and prisma combined with nextauth. But what I really want to achieve is having a separated backend with graphql so that all the nextauth handling is done in the backend instead of the frontend. Also from what I researched, users can see the api endpoints and easily call them and for example follow a user by just having their ids.

**Reply by faustinozanetto** [reply]
Maybe I would have to add some kind of encrypted key in the api request so that only me can call the api endpoints, what do you think?

**Reply by ryands17** [reply]
> Also from what I researched, users can see the api endpoints and easily call them and for example follow a user by just having their ids.

One way to achieve this is using `uuid` as the `id` field instead of an `int` so that the user cannot guess the id. This is the recommended approach as number based id's are easily guessable.

**Reply by ryands17** [reply]
> Maybe I would have to add some kind of encrypted key in the api request so that only me can call the api endpoints, what do you think?

Authentication of any sort wouldn't help here as the user making calls will already be authenticated. Best to make the id non-guessable.

**Reply by faustinozanetto** [reply]
I am already using uuid for user ids, the thing is: I should´t show the uuid in any part of the web app, right? Thanks

**Reply by ryands17** [reply]
In this case the user can follow anyone whose `id` they have. I don't think there's a way to prevent that as long as the user is authenticated and has valid user id's that they can follow.

**Reply by faustinozanetto** [reply]
I ended up doing this, don´t really know how efficient it is though.
```typescript
/* eslint-disable import/no-anonymous-default-export */
import { NextApiRequest, NextApiResponse } from 'next';
import prisma from '@lib/prisma/prisma';
import { getSession } from 'next-auth/client';

export default async (req: NextApiRequest, res: NextApiResponse) => {
  try {
    // Getting current user id and the user to follow id from the req body.
    const { id, targetId } = req.body;
    // Getting the session.
    const session = await getSession({ req });
    // If session is invalid we are not logged in.
    if (!session) {
      res.status(401);
      res.json({ error: 'Not signed in' });
    } else {
      // Find the signed in user with the session email
      const signedUser = await prisma.user.findUnique({
        where: {
          email: session.user?.email!,
        },
      });
      const userFromId = await prisma.user.findUnique({
        where: {
          id: id,
        },
      });
      // If ids do not match the current logged in user is not the same as the one from the request id.
      if (signedUser?.id !== userFromId?.id) {
        res.status(401);
        res.json({ error: 'Unauthorized' });
      } else {
        // Creating an userOnUser with the ids.
        const followed = await prisma.userOnUser.create({
          data: {
            parentId: id,
            childId: targetId,
          },
        });
        // Returning the result
        res.status(200);
        res.json(followed);
      }
    }
  } catch (error) {
    // Error handling.
    console.error(error);
    res.status(500);
    res.json({ error: 'Sorry unable to save sighting to database' });
  }
};
```

**Reply by ryands17** [reply]
This isn't needed:

```ts
const userFromId = await prisma.user.findUnique({
        where: {
          id: id,
        },
      });
```

You could just do:

```ts
if (signedUser?.id !== id) {
}
```

**Reply by faustinozanetto** [reply]
Yep I realised afterwards. So basically I think that now if anyone wants to call the endpoint from outside they need to be signed in with the account linked to the id they are using.

**Reply by faustinozanetto** [reply]
Thank you so much for helping me out!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
